miden-crypto/src/merkle/smt/full/large/mod.rs

```diff
@@ -0,0 +1,1112 @@
+                // default value
+                self.remove_inner_node(index);
+            } else {
+                self.insert_inner_node(index, InnerNode { left, right });
+            }
+        }
+        self.set_root(node_hash);
+    }
+
+    fn key_to_leaf_index(key: &RpoDigest) -> LeafIndex<SMT_DEPTH> {
+        let most_significant_felt = key[3];
+        LeafIndex::new_max_depth(most_significant_felt.as_int())
+    }
+
+    fn path_and_leaf_to_opening(path: MerklePath, leaf: SmtLeaf) -> SmtProof {
+        SmtProof::new_unchecked(path, leaf)
+    }
+}
+
+impl<S: SmtStorage + 'static> PartialEq for LargeSmt<S> {
+    fn eq(&self, other: &Self) -> bool {
+        self.root == other.root
+            && self.num_leaves() == other.num_leaves()
+            && self.num_entries() == other.num_entries()
+    }
+}
+
+impl<S: SmtStorage + 'static> Eq for LargeSmt<S> {}
+
+fn to_memory_index(index: &NodeIndex) -> usize {
+    debug_assert!(index.depth() <= IN_MEMORY_DEPTH);
+    debug_assert!(index.value() < (1 << index.depth()));
+    ((1usize << index.depth()) - 1) + index.value() as usize
+}
+
+#[allow(dead_code)]
+fn flat_idx_to_node_index(flat_idx: usize) -> NodeIndex {
+    let depth = (flat_idx + 1).ilog2() as u8;
+    let nodes_before_this_depth = (1usize << depth) - 1;
+    let value_at_depth = (flat_idx - nodes_before_this_depth) as u64;
+    NodeIndex::new_unchecked(depth, value_at_depth)
+}
+
+impl<S: SmtStorage + 'static> Clone for LargeSmt<S> {
+    fn clone(&self) -> Self {
+        Self {
+            root: self.root,
+            storage: Arc::clone(&self.storage),
+            in_memory_nodes: self.in_memory_nodes.clone(),
+            in_memory_count: self.in_memory_count,
+        }
+    }
+}
+
+// ITERATORS
+// ================================================================================================
+
+enum InnerNodeIteratorState<'a> {
+    InMemory {
+        current_index: usize,
+        large_smt_in_memory_nodes: &'a Vec<Option<InnerNode>>,
+    },
+    Subtree {
+        subtree_iter: Box<dyn Iterator<Item = Subtree> + 'a>,
+        current_subtree_node_iter: Option<Box<dyn Iterator<Item = InnerNodeInfo> + 'a>>,
+    },
+    Done,
+}
+
+pub struct LargeSmtInnerNodeIterator<'a, S: SmtStorage + 'static> {
+    large_smt: &'a LargeSmt<S>,
+    state: InnerNodeIteratorState<'a>,
+}
+
+impl<'a, S: SmtStorage + 'static> LargeSmtInnerNodeIterator<'a, S> {
+    fn new(large_smt: &'a LargeSmt<S>) -> Self {
+        // in-memory nodes should never be empty
+        Self {
+            large_smt,
+            state: InnerNodeIteratorState::InMemory {
+                current_index: 0,
+                large_smt_in_memory_nodes: &large_smt.in_memory_nodes,
+            },
+        }
+    }
+}
+
+impl<S: SmtStorage + 'static> Iterator for LargeSmtInnerNodeIterator<'_, S> {
+    type Item = InnerNodeInfo;
+
+    fn next(&mut self) -> Option<Self::Item> {
```

Comment:
This function is difficult to follow. There are multiple nested loops, iterators and indices are advanced in multiple places, and it's not clear what the expected control flow is. Can any of this be split out into helper functions, or organized into helper iterators/iterator adapters? It kind of seems to me like there's three different "effective iterators" this transitions between? If this can't really be split out or re-organized then it should be much more commented to clarify the expected flow.